### ISSUE TITLE
[Ch 21][Notebook][Major] "메인은 MLM 1 epoch" 전제가 부록 3곳·본편 FAQ Q3 에 잔존 — 메인은 3 epoch

### ISSUE THREAD

**Issue by yum-yeom** [yoon-gu/neuqes-101#129]
## 대상

- **챕터**: Ch 21
- **산출물**: Notebook 부록 (`appendix_compute_budget.ipynb` 3곳) + 본편 FAQ Q3 (`21_en_bert_classify.ipynb`)
- **기준 커밋**: `c16215e`
- **환경**: 정적 검토 + git 이력 확인

## 문제 위치

**결함 위치**

| 파일 | 위치 | 링크 | 내용 |
|---|---|---|---|
| `21_en_bert_classify/appendix_compute_budget.ipynb` | cell[5] — 데이터·모델 hyperparams 산문 | [L37](https://github.com/yoon-gu/neuqes-101/blob/c16215e/21_en_bert_classify/appendix_compute_budget.ipynb?plain=1#L37) | '메인은 1 epoch 으로 줄였지만' |
| `21_en_bert_classify/appendix_compute_budget.ipynb` | cell[13] — `## 2. 🅰️ Setup A` 도입 | [L95](https://github.com/yoon-gu/neuqes-101/blob/c16215e/21_en_bert_classify/appendix_compute_budget.ipynb?plain=1#L95) | 'MLM 만 1 → 3 epoch 로 늘려' |
| `21_en_bert_classify/appendix_compute_budget.ipynb` | cell[6] — 셋업 코드 주석 | [L43](https://github.com/yoon-gu/neuqes-101/blob/c16215e/21_en_bert_classify/appendix_compute_budget.ipynb?plain=1#L43) | '메인은 1 epoch — 분류 시간 확보용' |
| `21_en_bert_classify/21_en_bert_classify.ipynb` | cell[47] — FAQ Q3 코드 예시 | [L333](https://github.com/yoon-gu/neuqes-101/blob/c16215e/21_en_bert_classify/21_en_bert_classify.ipynb?plain=1#L333) | `MLM_EPOCHS = 3   # 1 -> 3` |

**반증**

| 파일 | 위치 | 링크 | 내용 |
|---|---|---|---|
| `21_en_bert_classify/21_en_bert_classify.ipynb` | cell[22] (실행본 코드셀 [11]) — MLM TrainingArguments | [L155](https://github.com/yoon-gu/neuqes-101/blob/c16215e/21_en_bert_classify/21_en_bert_classify.ipynb?plain=1#L155) | `MLM_EPOCHS = 3` |
## 실제 결과

메인 노트북의 MLM 은 **3 epoch** 입니다 (`MLM_EPOCHS = 3`, `8f41c67` 에서 1 → 3). 그런데 부록은 여전히 "메인은 1 epoch" 을 전제로 쓰여 있습니다.

| 자리 | 현재 표기 |
|---|---|
| 부록 §데이터·모델 hyperparams | *"메인 챕터는 분류 fine-tune 시간 확보를 위해 MLM 을 1 epoch 으로 줄였지만, 부록은 fair-compute 비교가 핵심이라 **MLM 1 → 3 epoch** 로 늘려"* |
| 부록 §2 Setup A 도입 | *"메인 챕터의 핵심 셋업을 **MLM 만 1 → 3 epoch** 로 늘려 재현합니다 — 1 epoch 으로는 본체 표상이 덜 정렬돼서"* |
| 부록 셋업 코드 주석 | *"메인은 1 epoch — 분류 시간 확보용. 부록은 사전학습 효과를 충분히 보기 위해 3 epoch."* |
| 본편 FAQ Q3 (개선 안내) | ```MLM_EPOCHS = 3   # 1 -> 3``` |

두 가지 문제가 있습니다.

1. **부록의 존재 이유가 성립하지 않습니다.** 메인도 부록도 MLM 3 epoch 이라, "부록은 메인보다 사전학습을 더 충분히 한다" 는 설계 근거가 사라졌습니다. 부록의 A 셋업은 메인과 같은 조건(데이터 크기만 2K/400 로 축소)입니다.
2. **본편 FAQ Q3 의 개선 안내가 무효입니다.** *"더 많은 사전학습으로 격차 줄이기"* 로 `MLM_EPOCHS = 3  # 1 -> 3` 을 제시하는데, 챕터가 이미 3 입니다. 학습자가 "이미 적용된 값" 을 개선책으로 시도하게 됩니다.

원인은 [#70](https://github.com/yoon-gu/neuqes-101/issues/70) 과 같습니다 — `8f41c67` 에서 코드가 바뀌었는데 산문이 따라가지 않았습니다.

## 기대 결과

부록·FAQ 의 서술이 현재 `MLM_EPOCHS` 값과 맞아야 하고, 부록이 메인과 무엇이 다른지가 정확히 적혀야 합니다.

## 수정 제안

- 부록 3곳의 "메인은 1 epoch" 을 걷어내고, 부록이 메인과 실제로 다른 점(분류·MLM 데이터 2K/400 로 축소, B/C 셋업 추가)만 남기기.
- 본편 FAQ Q3 의 개선 안내는 아직 적용되지 않은 축으로 교체 (예: `N_MLM_TRAIN` 을 2000 → 20000).
- 부록 실행(별건 이슈)과 함께 처리하면 한 번에 검증됩니다.

## 재현 절차

```bash
grep -o '"MLM_EPOCHS = [0-9]"' 21_en_bert_classify/21_en_bert_classify.ipynb
grep -c '1 epoch' 21_en_bert_classify/appendix_compute_budget.ipynb
git log --oneline -1 8f41c67
```

## 심각도

**Major** — 부록의 설계 근거와 본편의 개선 안내가 둘 다 틀린 전제 위에 있습니다.

## 담당

- 등록자·재검수자: @yum-yeom (페어 C)
- 담당: @fluentmin

## 연결

- 챕터 검수 이슈: [#45](https://github.com/yoon-gu/neuqes-101/issues/45)
- 같은 뿌리: [#70](https://github.com/yoon-gu/neuqes-101/issues/70) (`8f41c67` 이후 산문 미갱신)

**Comment by yum-yeom** [comment]
## 수정했습니다 — 부록 5곳 + 본편 FAQ Q3

브랜치 `fix/ch21-review-2026-09` · 커밋 [`b3a5045`](https://github.com/yoon-gu/neuqes-101/commit/b3a5045)

`8f41c67` 에서 메인이 MLM 1 → 3 epoch 으로 바뀐 뒤 산문만 남아 있던 자리입니다. 이슈 본문에 적은 4곳 외에 부록 첫 셀에서 2곳을 더 찾아 함께 고쳤습니다 (총 6곳).

### 부록 — "메인보다 더 충분히" 라는 설계 근거 자체가 사라진 상태였습니다

```diff
- ### 데이터·모델 hyperparams — 메인과 통일, MLM 은 더 충분히
- 메인 챕터는 *분류 fine-tune 시간 확보* 를 위해 MLM 을 1 epoch 으로 줄였지만, 부록은 fair-compute 비교가 핵심이라 **MLM 1 → 3 epoch** 로 늘려 …
+ ### 데이터·모델 hyperparams — 메인과 통일, 데이터만 축소
+ MLM 은 메인 챕터와 **같은 3 epoch** 입니다 — 부록이 메인과 다른 점은 *데이터 크기* 와 *셋업 B·C 추가* 뿐입니다. …
```

```diff
- | 🅰️ **A** | MLM 3 epoch | 2 epoch | 메인의 1 → 3 epoch (일반 도메인 사전학습 효과를 충분히, 기준선) |
+ | 🅰️ **A** | MLM 3 epoch | 2 epoch | 메인 챕터와 같은 3 epoch (일반 도메인 사전학습 기준선) |
```

- 첫 셀 "**메인 챕터와의 관계**" 문단 — 같은 취지로 다시 썼습니다
- §2 Setup A 도입 — "*MLM 만 1 → 3 epoch* 로 늘려 재현" → "메인의 핵심 셋업(MLM 3 epoch + 분류 2 epoch)을 *더 작은 데이터* 로 재현"
- 셋업 코드 주석 2줄 — "메인은 1 epoch — 분류 시간 확보용" → "메인 챕터와 같은 3 epoch — 부록이 다른 건 데이터 크기와 B·C 셋업뿐"

### 본편 FAQ Q3 — 이미 적용된 값을 개선책으로 안내하던 자리

```diff
  # 더 많은 사전학습으로 격차 줄이기 (T4 30분 룰 안에서)
- MLM_EPOCHS = 3                     # 1 -> 3
- # 또는 데이터 늘리기 — N_MLM_TRAIN 만 늘려도 효과 큼
+ # epoch 은 이 챕터가 이미 3 입니다 (MLM_EPOCHS = 3) — 다음으로 늘릴 축은 *데이터 양*
```

이어지는 "`N_MLM_TRAIN = 20000` + `MLM_EPOCHS = 1` 정도가 T4 30분 룰 안에서 최대치" 문장도 실측(2K × 3 epoch = 13초) 기준으로 다시 썼습니다 — #127 과 겹치는 자리라 함께 처리했습니다.

### 확인

```
$ grep -c '1 → 3 epoch\|1 epoch 로 줄였\|1 epoch 으로 줄였' 21_en_bert_classify/*.ipynb executed/21_en_bert_classify.ipynb
0
```

부록은 실행본이 없어 대조할 출력이 없습니다 (#128). 본편은 코드 변경이 없어 재실행하지 않았고 실행본 마크다운만 동기화했습니다([`9c32ad2`](https://github.com/yoon-gu/neuqes-101/commit/9c32ad2)).

@fluentmin 확인 후 닫아 주세요.

**Comment by yum-yeom** [comment]
## 정정 — 기준 브랜치와 #123 상태

등록할 때 쓴 기준이 틀렸습니다. 본문을 조용히 고치지 않고 여기 적습니다.

**1. 기준 브랜치·커밋**

`fix/ch21-review-2026-09` @ `c16215e` 로 적었는데, 팀이 통일한 Ch 21 브랜치는 **`ch-21/review-2026-09`** (Ch 20 의 `ch-20/review-2026-09` 와 같은 규칙) 이고 커밋 2개가 더 있습니다.

- `428f2c2` — CLAUDE.md 에 검수 기준 논의 #63 합의 4건 반영
- `8e65def` — Ch 21 모델 random init 시드 (#123)

**이 이슈의 결함은 새 브랜치(`8e65def`)에서도 그대로 재현됨을 확인했습니다.** 지적 내용에는 영향이 없습니다.

**2. #123 상태**

본문·앞선 댓글에서 "본편은 시드가 걸려 있지 않다" 고 적었는데, 새 브랜치에서는 `8e65def` 로 **이미 걸려 있습니다** (셀 15·27 모델 생성 직전 `set_seed(SEED)`). 제가 옛 브랜치만 보고 쓴 것입니다.

다만 그 커밋이 코드셀 3개(7·15·27)를 바꾼 뒤 **재실행이 되지 않아** 실행본은 8/13 판 그대로입니다 (`executed_from.executed_at = 2026-08-13`, `source_sha256` 불일치). #123 의 Ch 21 몫은 **재실행이 남아 있습니다.**

**3. 수정 커밋 위치**

앞 댓글의 `b3a5045` · `9c32ad2` 는 옛 브랜치 커밋입니다. 새 브랜치로 옮겨 다시 올렸습니다.

- [`ec8e175`](https://github.com/yoon-gu/neuqes-101/commit/ec8e175) — 서술 수정 (내용 동일)
- [`4bae94f`](https://github.com/yoon-gu/neuqes-101/commit/4bae94f) — 실행본 마크다운 동기화. `source_sha256` 은 실제 실행된 소스(`c16215e`) 값을 그대로 두고 `executed_from.note` 에 사유를 남겼습니다 — 지금의 해시 불일치가 **#123 재실행 필요** 를 뜻하도록.


**4. 본문에서 빠졌던 두 필드**

## 전파 범위

- 부록 5곳은 변환 경로 밖이라 `pages/`·tex 전파 없음
- 본편 FAQ Q3 는 전파됨 — `pages/21-en_bert_classify-wrapup.md:69`, `book/chapters/ch21_en_bert_classify.tex:1858`

## 재실행 필요 여부

**불필요** — 마크다운 셀과 부록 코드 *주석* 만 고쳤습니다 (실행 동작 변경 없음).

**Comment by fluentmin** [comment]
확인했습니다 (@fluentmin, `ch-21/review-2026-09` @ `4bae94f`).

"메인은 1 epoch" 전제가 노트북 두 개와 실행본 어디에도 남아 있지 않습니다.

```
$ grep -c '1 → 3 epoch\|1 epoch 로 줄였\|1 epoch 으로 줄였\|메인은 1 epoch' \
    21_en_bert_classify/21_en_bert_classify.ipynb \
    21_en_bert_classify/appendix_compute_budget.ipynb \
    executed/21_en_bert_classify.ipynb
0 / 0 / 0
```

부록 셋업 코드의 `MLM_EPOCHS = 3` 과 주석("메인 챕터와 같은 3 epoch — 부록이 다른 건 데이터 크기와 B·C 셋업뿐"), 셋업 표 A 행, 첫 셀 "메인 챕터와의 관계" 문단이 모두 같은 이야기를 합니다. 부록의 존재 이유를 *"메인보다 사전학습을 더 충분히"* 에서 *"데이터 크기 축소 + B·C 셋업 추가"* 로 다시 잡은 게 이 이슈의 핵심이었는데, 그 방향으로 정리됐습니다.

본편 FAQ Q3 도 확인했습니다. 이미 3 인 값을 개선책으로 안내하던 자리가 `N_MLM_TRAIN` 2000 → 20000 으로 바뀌었고, 이슈 본문에서 지적하지 않았던 이어지는 문장("`MLM_EPOCHS = 1` 정도가 30분 룰 최대치")까지 실측 기준으로 다시 쓰셨습니다.

### 부록을 실제로 돌려 교차검증했습니다

정적 대조만으로 닫기 전에 부록을 Colab T4 에서 끝까지 실행해(`status: ok`, 141.1초) 산문이 아니라 *실제 동작* 과 맞는지 확인했습니다.

```
[198/198 00:12, Epoch 3/3]        ← MLM 진행표
[A] MLM pretraining: 0.23 min  (mean train loss: 7.6115)

setup               pretraining   fine-tune epochs   total compute (min)
A (MLM + cls)       MLM 3 epoch                  2                  0.34
```

§5 비교표가 A 셋업을 `MLM 3 epoch` 으로 찍고, 진행표도 `Epoch 3/3` 으로 3 epoch 을 돕니다. "메인 챕터와 같은 3 epoch" 이라는 수정 후 서술이 코드·출력과 모두 일치합니다 — 고친 산문이 실제와 맞는 것까지 확인했습니다.

부록 실행본 자체는 아직 커밋하지 않았습니다. 이번 실행에서 A/B/C 결과가 부록이 예상한 방향과 다르게 나와, 그 논의는 #128 에서 이어갑니다. 다만 이 이슈(= `MLM_EPOCHS` 값과 산문의 일치)의 판정에는 영향이 없습니다.

`pages/21-en_bert_classify-wrapup.md:69` · `book/chapters/ch21_en_bert_classify.tex:1858` 재변환은 #45 에서 이어가고, 이 이슈는 닫습니다.
